dll/win32/wdmaud.drv/legacy.c

```diff
@@ -1,89 +1,306 @@
+    PWDMAUD_DEVICE_INFO DeviceInfo = (PWDMAUD_DEVICE_INFO)Parameter;
+    PWAVEHDR_EXTENSION HeaderExtension;
+    PWAVEHDR WaveHeader;
+#ifndef USE_MMIXER_LIB
     MMRESULT Result;
-    WDMAUD_DEVICE_INFO DeviceInfo;
-    PSOUND_DEVICE_INSTANCE Instance = (PSOUND_DEVICE_INSTANCE)Parameter;
+#endif
+
+    /* Enter critical section */
+    EnterCriticalSection(&WdmAudCriticalSection);
+
+    while(TRUE)
+    {
+        DPRINT("WaveQueue %p\n", DeviceInfo->DeviceState.WaveQueue);
+        if (DeviceInfo->DeviceState.WaveQueue)
+        {
+            DPRINT("1\n");
+            WaveHeader = DeviceInfo->DeviceState.WaveQueue;
+            if (WaveHeader)
+            {
+                DPRINT("2\n");
+                HeaderExtension = (PWAVEHDR_EXTENSION)WaveHeader->reserved;
+                if (HeaderExtension)
+                {
+                    DPRINT("3\n");
+
+                    /* Leave critical section */
+                    LeaveCriticalSection(&WdmAudCriticalSection);
+
+                    /* Wait for I/O complete */
+                    WaitForSingleObject(HeaderExtension->Overlapped->hEvent, INFINITE);
+
+                    /* Enter critical section */
+                    EnterCriticalSection(&WdmAudCriticalSection);
+
+                    DPRINT("4\n");
+                    if (DeviceInfo)
+                    {
+                        /* Complete current header */
+                        CompleteWaveHeader(DeviceInfo);
+                    }
+                    else
+                    {
+                        DPRINT1("Invalid device info data %p\n", DeviceInfo);
+                        break;
+                    }
+                }
+            }
+            /* Leave critical section */
+            LeaveCriticalSection(&WdmAudCriticalSection);
+        }
+        else
+        {
+            DPRINT("6\n");
+            if (DeviceInfo->DeviceState.bStart)
+            {
+                DPRINT("7\n");
+#ifdef USE_MMIXER_LIB
+
+                /* make sure the pin is stopped */
+                MMixerSetWaveStatus(&MixerContext, DeviceInfo->hDevice, KSSTATE_PAUSE);
+                MMixerSetWaveStatus(&MixerContext, DeviceInfo->hDevice, KSSTATE_ACQUIRE);
+                MMixerSetWaveStatus(&MixerContext, DeviceInfo->hDevice, KSSTATE_STOP);
+
+#else
+                /* Stop streaming if no more data to stream */
+                Result = WdmAudIoControl(DeviceInfo,
+                                         0,
+                                         NULL,
+                                         DeviceInfo->DeviceType == WAVE_OUT_DEVICE_TYPE ?
+                                         IOCTL_PAUSE_PLAYBACK : IOCTL_PAUSE_CAPTURE);
+                if (!MMSUCCESS(Result))
+                {
+                    DPRINT1("Call to %ls failed with %d\n",
+                            DeviceInfo->DeviceType == WAVE_OUT_DEVICE_TYPE ?
+                            L"IOCTL_PAUSE_PLAYBACK" : L"IOCTL_PAUSE_CAPTURE",
```

Comment:
Using wide strings here isn't ideal, DbgPrint is natively ansi

Reply:
Okay, changed all strings to ANSI here and in all other places. :slightly_smiling_face: